extensions/amp-smartlinks/0.1/linkmate.js

```diff
@@ -0,0 +1,180 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {deepEquals} from '../../../src/json';
+import {dict} from '../../../src/utils/object';
+
+import {ENDPOINTS} from './constants';
+import {TwoStepsResponse} from
+  '../../amp-skimlinks/0.1/link-rewriter/two-steps-response';
+import {getData} from '../../../src/event-helper';
+
+
+export class Linkmate {
+  /**
+   * @param {?../../../src/service/ampdoc-impl.AmpDoc} ampDoc
+   * @param {?../../../src/service/xhr-impl.Xhr} xhr
+   * @param {?Object} linkmateOptions
+   */
+  constructor(ampDoc, xhr, linkmateOptions) {
+    /** @private {?../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampDoc_ = ampDoc;
+
+    /** @private {?../../../src/service/xhr-impl.Xhr} */
+    this.xhr_ = xhr;
+
+    /** @private {?boolean} */
+    this.requestExclusiveLinks_ = linkmateOptions.exclusiveLinks;
+
+    /** @private {?number} */
+    this.publisherID_ = linkmateOptions.publisherID;
+
+    /** @private {?string} */
+    this.linkAttribute_ = linkmateOptions.linkAttribute;
+
+    /** @private {!Document|!ShadowRoot} */
+    this.rootNode_ = this.ampDoc_.getRootNode();
+
+    /** @private {?Array<!HTMLElement>} */
+    this.anchorList_ = null;
+
+    /** @private {?Array<JsonObject>}*/
+    this.linkmateResponse_ = null;
+  }
+
+  /**
+   * Callback used by LinkRewriter. Whenever there is a change in the anchors
+   * on the page we want make a new API call.
+   * @param {!Array<!HTMLElement>} anchorList
+   * @return {!../../amp-skimlinks/0.1/link-rewriter/two-steps-response.TwoStepsResponse}
+   * @public
+   */
+  runLinkmate(anchorList) {
+    // If we already have an API response and the anchor list has
+    // changed since last API call then map any new anchors to existing
+    // API response
+    let syncMappedLinks = null;
+    if (this.linkmateResponse_ && this.anchorList_ &&
+      !deepEquals(this.anchorList_, anchorList)) {
+      syncMappedLinks = this.mapLinks_();
+    }
+
+    // If we don't have an API response or the anchor list has changed since
+    // last API call then build a new payload and post to API
+    if (!this.linkmateResponse_ ||
+      (this.anchorList_ && !deepEquals(this.anchorList_, anchorList))) {
```

Comment:
From reading the `link-rewriter` implementation this function will be called anytime new links are written to the DOM. I implemented this so that this code would fit into `link-rewriter` nicely. If that isn't actually the case I can simplify this like you suggested.

Reply:
Ah! You're right, forgot about DOM change. Thank you for clarifying